File: medium/Subsets.py
class Solution:
    def __init__(self):
        self.answer = [[]]
        
        
    def next_number(self, nums, answer, cnt):
        if len(answer) == cnt:
            self.answer.append(answer)
        for index in range(len(nums)):
            self.next_number(nums[index+1:], answer+[nums[index]], cnt)
        
    def subsets(self, nums: List[int]) -> List[List[int]]:
        for r in range(1, len(nums)):
            for index in range(len(nums)):
                self.next_number(nums[index+1:], [nums[index]], r)
        self.answer.append(nums)
        return self.answer

File: cpp/medium/Subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> answer;
        vector<int> combination;
        dfs(nums, answer, combination);
        return answer;
    }
    void dfs(vector<int>& nums, vector<vector<int>>& answer, vector<int>& combination) {
        answer.push_back(combination);
        for(int i = 0; i < nums.size(); ++i) {
            combination.push_back(nums[i]);
            vector<int> sub_nums(nums.begin() + i + 1, nums.end());
            dfs(sub_nums, answer, combination);
            combination.pop_back();
        }
    }
};
